flink-runtime/src/test/java/org/apache/flink/runtime/state/filesystem/FileStateHandleTest.java

```diff
@@ -57,27 +59,97 @@ public void testDisposeDeletesFile() throws Exception {
      */
     @Test
     public void testDisposeDoesNotDeleteParentDirectory() throws Exception {
-        File parentDir = tempFolder.newFolder();
-        assertTrue(parentDir.exists());
+        final Path p = new Path(TEST_SCHEME + "://path/with/parent");
+        final List<Path> pathsToDelete = new ArrayList<>();
 
-        File file = new File(parentDir, "test");
-        writeTestData(file);
-        assertTrue(file.exists());
+        initializeFileSystem(
+                (path, ignoredRecursionMarker) -> {
+                    pathsToDelete.add(path);
+                    return true;
+                });
 
-        FileStateHandle handle = new FileStateHandle(Path.fromLocalFile(file), file.length());
+        final FileStateHandle handle = new FileStateHandle(p, 42);
         handle.discardState();
-        assertFalse(file.exists());
-        assertTrue(parentDir.exists());
+        assertThat(pathsToDelete)
+                .as(
+                        "Only one delete call should have happened on the actual path but not the parent.")
+                .singleElement()
+                .isEqualTo(p);
```

Comment:
TBH, the older version seems more readable to me. It also doesn't need to be updated if `FileStateHandle` internals are changed.
But I'm fine with this version as well.

Reply:
I agree. But the older version didn't enable us to simulate errors. So, I sticked to it for consistency reasons.

Reply:
I see. Thanks for clarifying.